### PR TITLE
Export Schedule To Image

### LINKED ISSUE #329: Feature Request — Add button to export schedule grid as an image
**Job**
Currently we allow users to export schedules in ics format and the expectation for images is that they can screenshot it. But it would be helpful if we provide an option for users to export the schedule as an image

**Describe a possible solution**
Another button next to `export as ics` with `export as image` that exports the schedule grid as an image
Or change that button to a dropdown menu

**Context**

### PR DESCRIPTION
**Pull request by venkatrsrinivas**
Hello Everyone!
This Is My Pull Request Concerning The Brand New Feature To Export Schedules As Images, Specifically PNG Files. 
On Top Of This, This Pull Request Extends Functionality By Creating A Drop-Down To Support Future Exportation Types. 
Below You Will Find All Important Details:

**Issue**

This Pull Request Should Ideally Close #329 . 

**Database Changes/Migrations**

Not Applicable.

**Full Testing Procedure**

*Sample Example(s) To Test Functionality:*
> 1. Run Spring 2021 w/ YACS + Select Some Courses.
> 2. Navigate To The Export Data Drop Down Menu.
> 3. Select Export To ICS + Verify Existing Functionality. 
> 4. Select Export To Image + Verify New Functionality. 
> 5. Should Immediately Receive Download. 

NOTE: I Also Added ERROR Messages To Alert Client If No Courses Are Selected. 

**All Relevant Screenshots**

Before: 

Screen Shot 2020-11-18 at 18 55 28

After: 

Screen Shot 2020-11-18 at 18 52 14
Screen Shot 2020-11-18 at 18 52 21
Screen Shot 2020-11-18 at 18 53 23

### CODE REVIEW

**Comment by marchdan** [comment]
@venkatrsrinivas Looks good! A couple of suggestions:
1. Have you thought about if this should conform to the dark mode status?
2. Should the exported calendar list the registration CRNs like the on-site schedule does?
3. Total feature request, but I think the image should also have the name of the semester so people don't get confused if they are working to schedule multiple semesters (i.e Summer & Fall)

File: src/web/src/utils.js
Comment (jshom) [inline src/web/src/utils.js:320]: Let's name this like yacs-{{semester}}.png
I.e. yacs-spring-2021.png for this semester

This way people can more easily search for the images

**Comment by venkatrsrinivas** [comment]
@marchdan, @jshom 
1. I did think about this, but currently the Export Button Styling is the same in both Dark Mode On/Off, so I left that alone. 
2. That's a good point, I could try to see if that is possible, but as of right now, that is not being considered. 
3. Definitely, I agree. Will do following Jacob's Suggestion above.

**Comment by marchdan** [comment]
@venkatrsrinivas 
1. I meant for the actually exported image to follow dark mode protocols
2. I think this would be nice if you can easily add it, if not I would made a feature request issue for it after this is merged
3. That works for me

**Comment by venkatrsrinivas** [comment]
@marchdan
1. Oh, ... As of right now, it defaults to Dark Mode On for some reason, but I can fix that hopefully by just passing Dark Mode to the Function. 
I have the File Name Fixed now, so what is left is to get the CRN Data + Ideally Dark Mode.

**Comment by marchdan** [comment]
@venkatrsrinivas It shows as dark mode for you when exported? Mine shows as light mode when exported...

**Comment by venkatrsrinivas** [comment]
Yes I usually have YACS on Dark Mode so all of my Images are Dark Mode ... Strange.

**Comment by marchdan** [comment]
> @marchdan Yes I usually have YACS on Dark Mode so all of my Images are Dark Mode ... Strange.

Hmmm, I wonder if the module doing the export loads the setting on page load and doesn't change when dark mode is toggled? I think explicitly ensuring the exported image is in Dark mode or not should solve it.

**Comment by venkatrsrinivas** [comment]
@marchdan Yes definitely that should work. Just trying to see how to actually do that and set the Exported Image Background. Will keep you all posted.

**Comment by venkatrsrinivas** [comment]
Alright, I have it mostly working now with Dark Mode, except for reason, on Dark Mode, the Image Background is completely Black as opposed to our Dark Mode Color even though I set the Color as that from the Dark SCSS File.

File: src/web/src/utils.js
Comment (marchdan) [inline src/web/src/utils.js:320]: I don't think this needs to be as long a name, it confuses me when I download it, so it will probably confuse others. It's more important they see the semester than that they know its from yacs.n. I think even just "_semester__yacs-schedule.png" would be fine.

**Comment by marchdan** [comment]
> Alright, I have it mostly working now with Dark Mode, except for reason, on Dark Mode, the Image Background is completely Black as opposed to our Dark Mode Color even though I set the Color as that from the Dark SCSS File.

@venkatrsrinivas Could you use something like:
`element.classList.add("customClass");` and then style it all using dark.scss?

File: src/web/src/utils.js
Comment (venkatrsrinivas) [reply src/web/src/utils.js]: Sounds good to me. 
Done.

**Comment by venkatrsrinivas** [comment]
> > Alright, I have it mostly working now with Dark Mode, except for reason, on Dark Mode, the Image Background is completely Black as opposed to our Dark Mode Color even though I set the Color as that from the Dark SCSS File.
> 
> @venkatrsrinivas Could you use something like:
> `element.classList.add("customClass");` and then style it all using dark.scss?

Alright I will try that out. 
I'm not entirely sure why it is not working with the Current Implementation though since it is literally using the Dark.SCSS File already.

**Comment by venkatrsrinivas** [comment]
I was not sure how to get it working with element.classList.add("customClass") but I did figure out how to export the SCSS Variable so that I could have the Schedule store the isDarkModeOn Color to then be utilized by utils.js. 
Here were Links that really helped in case in the future anyone needs to Export SCSS Variables in JavaScript:
1. https://dev.to/pecus/how-to-share-sass-variables-with-javascript-code-in-vuejs-55p3
2. https://www.bluematador.com/blog/how-to-share-variables-between-js-and-sass
Now, the Back-Ground Color should be fine in the Exported Images.
I'll try to finalize this Feature by adding the CRN Data as well.

**Comment by venkatrsrinivas** [comment]
Alright, I think the CRN should be working as well now.
I had to make changes here and there, but I think everything is sound and should work as intended.
Kindly let me know and I will post Updated Screenshots.

**Comment by marchdan** [comment]
@venkatrsrinivas It's not exporting in dark mode for me:

Light mode export:
SPRING 2021-YACS-Schedule


Dark mode export:
SPRING 2021-YACS-Schedule (1)

I am using the https://45.56.102.81/ deploy link

**Comment by venkatrsrinivas** [comment]
@marchdan  
Hmm this is really strange.
I just replicated why you are seeing, but I learned something else.
When I open the Image on my Local Machine, this is what I see via Screenshot: 

Screen Shot 2020-11-22 at 00 39 50


However, when I upload the Image Files here to GitHub, this is what I get for Dark Mode: 
SPRING 2021-YACS-Schedule (6)

**Comment by marchdan** [comment]
> @marchdan  
> 
> Hmm this is really strange.
> 
> I just replicated why you are seeing, but I learned something else.
> 
> When I open the Image on my Local Machine, this is what I see via Screenshot: 
> 
> 
> 
> Screen Shot 2020-11-22 at 00 39 50
> 
> 
> 
> 
> 
> However, when I upload the Image Files here to GitHub, this is what I get for Dark Mode: 
> 
> SPRING 2021-YACS-Schedule (6)
> 
> 
> 
> 

Have you tried opening the dark mode image in a photo editor? When I opened mine in paint the background was totally transparent, but you will only see that in a photo editor. (I can try downloading on Linux tomorrow and see if I get the same results)

**Comment by venkatrsrinivas** [comment]
Alright sounds good thank you.
Personally, I opened the Image in Apple's Preview and it looks right, but for sure something is wrong.
Right not I'm just setting the bgcolor Attribute the same way the Old YACS.
The Color Value it's been set to is fine, but it seems that the Problem is in the fact it seems like that it's not enough to just to set that Attribute for the Image.

**Comment by marchdan** [comment]
@venkatrsrinivas The dark mode images are definitely exporting with a transparent background. The site you linked above has a dark background by default, so you can't tell. You need to open it in a photo editing software that shows transparent backgrounds different from the default background (I'm using GNU, or paint 3D)

File: src/web/src/components/Schedule.vue
Comment (marchdan) [inline src/web/src/components/Schedule.vue:13]: This tab should be removed as no actual changes were made to this file

**Review by marchdan — COMMENTED** [review]
It looks like you might to do something similar to what is listed in step 1 here: https://til.hashrocket.com/posts/sxbrscjuqu-share-scss-variables-with-javascript

In /src/web/vue.config.js

**Comment by venkatrsrinivas** [comment]
@marchdan 
Alright thank you. 
I'm trying to add that Step 1 in, it's just that I get an ERROR when trying to run YACS telling me so,ething similar to:
`ERROR  Invalid options in vue.config.js: "module" is not allowed` ...
I've been trying to see if there is something specific to Vue.JS that is causing this.

File: src/web/src/components/Schedule.vue
Comment (venkatrsrinivas) [reply src/web/src/components/Schedule.vue:13]: Yes definitely. I'm not even sure how that came about. Probably because I had the Image of just the Schedule but I moved the Code to include the CRN. Nevertheless, will be fixed.

**Comment by marchdan** [comment]
> @marchdan 
> 
> Alright thank you. 
> 
> I'm trying to add that Step 1 in, it's just that I get an ERROR when trying to run YACS telling me so,ething similar to:
> 
> `ERROR  Invalid options in vue.config.js: "module" is not allowed` ...
> 
> I've been trying to see if there is something specific to Vue.JS that is causing this.

@venkatrsrinivas It will for sure be different syntax than what is in the example I sent, but it should he similar.

**Comment by 123joshuawu** [comment]
I believe scss files should already be loaded, I think you need to do the export and import parts

**Comment by venkatrsrinivas** [comment]
@123joshuawu I already did the Export + Import Parts, which is actually why I thought it was working when I saw the Image on my Mac.

File: src/web/src/utils.js
Comment (123joshuawu) [inline src/web/src/utils.js:316]: instead of trying to get the background color of the schedule with an attribute, maybe we can just pass it in as a parameter to `exportScheduleToImage`?

Comment (venkatrsrinivas) [reply src/web/src/utils.js:316]: Hmm I can try to do that, but that should not be causing the Color to be Transparent though, right?

Comment (123joshuawu) [reply src/web/src/utils.js:316]: Well if the attribute name doesn't end up being `isDarkModeOn` for some reason, maybe getAttribute will return undefined

Comment (venkatrsrinivas) [reply src/web/src/utils.js:316]: Hmm but I have been doing Inspect Element and that is the Attribute Name.
But when I output it, it has NULL as opposed to the Color.

**Comment by 123joshuawu** [comment]
FALL 2020-YACS-Schedule-12
Seems like a number of weird stuff is happening and not sure why, but this is the best version I could produce.

File: src/web/package.json
Comment (123joshuawu) [inline src/web/package.json:20]: we still using these other packages?

**Review by 123joshuawu — APPROVED** [review]

File: src/web/package.json
Comment (venkatrsrinivas) [reply src/web/package.json:20]: I'm not too sure, but I only added that DOM-To-Image One.

**Comment by venkatrsrinivas** [comment]
Here are the Final Screenshots.
I've managed to fix the Background Color with help from Josh, but there is some Random Bug with Exporting SCSS Variables that we will need to look into later.
I will make another Issue for this, but obviously, this would not be Priority 1 since the Functionality is still there as intended.
It would be more so for Coding Style + Better Extensibility in the Future in case Dark Mode Colors change. 

Here is the Exported Image now: 
SPRING-2021-YACS-Schedule
SPRING-2021-YACS-Schedule (1)

**Comment by venkatrsrinivas** [comment]
@marchdan Since you are the most familiar with this Pull Request as you have helped me a lot, could you take a look whenever you have time and potentially approve the Pull Request if all is good? 

@123joshuawu Wanted the Pull Request merged in and I think it's ready as well.

**Review by jshom — APPROVED** [review]

**Pull request merged**
